More KATANA_ERROR

witchel opened the pull request:
- ~~Pass expected number of hosts to RDG to enable a better error message when a user opens a partitioned graph on the wrong number of hosts.~~
-  Finish support for UINT32 in ImportData
- KATANA_ERROR in PropertyGraph.cpp, and ArrowInterchange.h

roshandathathri commented on lonestar/CMakeLists.txt:
```diff
@@ -110,47 +114,76 @@ if(GALOIS_ENABLE_DIST)
-      set_tests_properties(run${suffix}
-        PROPERTIES ENVIRONMENT GALOIS_DO_NOT_BIND_THREADS=1 LABELS quick)
+    if(NOT X_NOT_QUICK)
+      set_tests_properties(verify${suffix} PROPERTIES LABELS quick)
```
Is run${suffix} also set to quick?

ddn0 replied:
So the idea behind cmake test fixtures is to add `FIXTURE_REQUIRED` tests whenever the corresponding "leaf" test needs to run. This avoids an issue where you ask to run a test directly (e.g., `ctest -R regex`) but cmake does not run its dependencies (e.g., `setup.*${suffix}`). Since `run${suffix}` is marked as a required fixture of `verify${suffix}`, if the latter runs, cmake will make sure the former does as well.

I did modify the PR after your comment to make `setup.*${suffix}` a fixture requirement of `run${suffix}` as well, so you can also select `run${suffix}` and `setup.*${suffix}` will automatically be included.

l-hoang commented:
FYI the result checker script has been modified in the Cypher PR to work with the output from the Cypher engine. I'll take a look at changing it to work after both PRs get merged, or do you want me to check now?

tylershunt commented on .github/workflows/cpp.yaml:
```diff
@@ -122,67 +99,51 @@ jobs:
         GRAPH_FILE="$HOME/.cache/graph/inputs.tar.gz"
         GRAPH_LOCATION="$HOME/.cache/graph/inputs"
-        INPUT_URL="http://iss.ices.utexas.edu/projects/galois/downloads/small_inputs_for_lonestar_test.tar.gz"
+        INPUT_URL="https://katana-ci-public.s3.us-east-1.amazonaws.com/inputs/katana-inputs-v1.tar.gz"
```
AFAICT: There doesn't seem to be a way to restrict access to private runners (as a mechanism to protect AWS credentials) on public repos. If we want CI to run on pull requests anyone is free to fork and run code on our private runner. The closest thing is a moderation feature that prevents pull requests from anyone but collaborators for 24 hours.

So this public bucket looks like the right thing for now.

ddn0 replied:
Yeah GitHub actions is pretty bare bones.

It's basically a race as to whether our needs outpace their delivery to the point where we switch to something else. I think we should aim for "either it is obvious and trivial to do with GitHub actions or add an item to our dream CI list (i.e., no hacky workarounds)".

Having better support for forks seems to be on the GitHub roadmap https://github.community/t/will-github-actions-support-pull-request-events-from-a-fork-to-a-private-base-repository/17471 but I wouldn't hold your breath.

We could switch to a model where contributors write to branches in Katana directory. It depends on how hacky you think that is and how much we want to ease external contributions (probably not that much).

On the bright side, GitHub actions don't do much so it should be easy to port them to another system :D

ddn0 commented:
> FYI the result checker script has been modified in the Cypher PR to work with the output from the Cypher engine. I'll take a look at changing it to work after both PRs get merged, or do you want me to check now?

It is probably easiest for whoever comes after (probably you) to update support in the codebase to accommodate new features.

witchel committed:
More KATANA_ERROR

witchel commented on libtsuba/include/tsuba/RDG.h:
```diff
@@ -31,6 +31,8 @@ struct KATANA_EXPORT RDGLoadOptions {
   std::optional<uint32_t> partition_id_to_load;
+  /// Client is trying to load RDG into expected_num_hosts machines
+  std::optional<uint32_t> expected_num_hosts;
   /// List of node properties that should be loaded
```
@scober, I assume this is not an abuse of your RDGLoadOptions class.

scober replied:
Is there a reason to not use `getSystemNetworkInterface().Num`?

witchel replied:
There is no reason.  Can I blame the fact that I don't like global variables?  Maybe it was the phase of the moon.  Regardless, more discussion on this issue in #957, but I'm eliminating this change.

ddn0 commented on libsupport/include/katana/ArrowInterchange.h:
```diff
@@ -91,8 +91,9 @@ UnmarshalVector(const std::shared_ptr<arrow::ChunkedArray>& source) {
-    return katana::ErrorCode::ArrowError;
+    return KATANA_ERROR(
+        katana::ErrorCode::ArrowError, "UnmarshalVector arrow error: {}",
+        r.ToString());
```
Just `r` instead of `r.ToString()` should work here too (I think).

We already get some source line information when creating an error. It might be better to rewrite the message itself to be a little less implementation specific:

```
KATANA_ERROR(...., "converting buffer to vector: {}", r)
```

witchel replied:
Yeah, less implementation-specific messages are a good idea.  That is a failure of imagination.  And I forget the arrow results have a string conversion.  Fixed.

ddn0 commented on libsupport/include/katana/ArrowInterchange.h:
```diff
@@ -149,13 +152,14 @@ katana::Result<void>
     if (!res) {
-      KATANA_LOG_DEBUG("UnmarshalVectorOfVectors arrow error: {}", res.error());
-      return res.error();
+      return res.error().WithContext("UnmarshalVectorOfVectors");
```
maybe `WithContext("converting chunk {}", i)`

witchel committed:
Load partitioned RDGs that lack some gluon mapping data

witchel commented:
Added some code in `RDG.cpp` to make loading partitioned RDGs backward compatible with Roshan's recent gluon changes.  I spoke with him about the changes and added him as a reviewer.

roshandathathri commented on libtsuba/src/RDG.cpp:
```diff
@@ -385,17 +385,25 @@ tsuba::RDG::DoMake(const katana::Uri& metadata_dir) {
+      set_local_to_user_id(local_to_global_id_->Slice(0));
+    } else if (local_to_user_id_->length() != local_to_global_id_->length()) {
+      if (local_to_global_id_ != nullptr) {
+        if (local_to_user_id_ == nullptr) {
```
Both of these checks are superfluous. If either of them are `nullptr`, then `->length()` above should crash. In this case, the constructor guarantees that they are not `nullptr`.

witchel replied:
Sorry, I blindly threw that test in there and you are right, the logic should be cleaned up.  Done.

roshandathathri commented on libtsuba/src/RDG.cpp:
```diff
@@ -622,16 +639,20 @@ tsuba::RDG::SetTopologyFile(const katana::Uri& new_top) {
+void
+tsuba::RDG::InitArrowVectors() {
   // Create an empty array, accessed by Distribution during loading
+  host_to_owned_global_ids_ = katana::EmptyChunkedArray(arrow::uint64(), 0);
```
I'm not sure how Distribution handles these empty arrays. And I'm surprised this is working. But it doesn't block this PR from being merged.

witchel replied:
In `LoadPartMetadata`, Distribution.cpp calls `Unmarshal*` routines passing in `graph_pg_.local_to_user_id()` and the like.  `ArrowInterchange.h:87` calls `source->length()` and crashes.  This path is only exercised if you load a partitioned graph.  I don't think any of our CI tests load a partitioned graph.

roshandathathri replied:
I got that. I meant after that, if the arrays have 0 length, I don't understand how it works.

witchel replied:
Oh, they just have to be initialized because the code assumes it.  The loading process creates new arrays, causing the initial arrays to become garbage.

roshandathathri replied:
But in the previously partitioned graphs (for backward compatibility), it would not load anything for those arrays. That's what I'm confused about. Is that not a concern here?

witchel replied:
Oh, that is your department :)  I don't understand the semantics of these arrays well enough to know if I can live without them.  If not, just tell me that your check in was a breaking change and I'll repartition my graphs.

roshandathathri replied:
Hehe. @scober how does `DistGraphView` handle this?

scober replied:
I'm not sure I understand the question. `DistGraphView` defers almost all work to the underlying `Distribution`. So a malformed `Distribution` will result in bad/undefined behavior from the `DistGraphView`. Does that address your question?

roshandathathri replied:
Yes, that's what I was expecting. Which is why I'm surprised that it works for hosts > 1. @witchel Is the Partitioner not called on these loaded graphs? That is, is it going through UpdateGraphView directly?

witchel replied:
So far as I know, the partitioner is not called when we load a partitioned graph.

witchel replied:
I can confirm that I'm getting Partition errors whenever I run updates on graphs partitioned before Roshan's recent checkin.  There is a debug assert that fires on Partition.cpp:1026 in the debug build, I think it seg faults in the release build.  I'm repartitioning all my graphs.

scober replied:
Is there a convenient place to bail out with a helpful error message if we are trying to partition an outdated RDG? Or do we want to just assert/seg fault and tell people they have to update their RDGs?

witchel replied:
Excellent question.  I'm out of juice for this issue in this PR.  What I think is that old partitioned graphs should be destroyed, otherwise some things might work, but others might not.  Maybe only updates are broken, in which case no one but me cares, but I don't know.  This is a consequence of Roshan's #940, so far as I understand, which I would document as a breaking change for partitioned graphs.

For the sake of a single summary, I also see messy exits (misleading error message or segv) when loading a graph partitioned for N hosts on m&lt;N and m&gt;N hosts.  This PR does not affect that.

roshandathathri replied:
Thanks, Emmet.

roshandathathri reviewed (approved):
LGTM.

Some minor comments below.

scober commented on libtsuba/src/RDG.cpp:
```diff
@@ -385,14 +385,15 @@ tsuba::RDG::DoMake(const katana::Uri& metadata_dir) {
+      // NB: this is a zero-copy slice, so the underlying data is shared
+      set_local_to_user_id(local_to_global_id_->Slice(0));
+    } else if (local_to_user_id_->length() != local_to_global_id_->length()) {
+      if (local_to_user_id_->length() != local_to_global_id_->length()) {
```
This if statement seems redundant.

witchel replied:
I'm out of my depth here.  I added what Roshan told me to add, and it seems rational to only set local_to_user_id if it wasn't set by the RDG.  But @roshandathathri, can you comment?  

In general, in this file, I see these chunked arrow arrays being tested for nullptr.  The lifecycles of RDGs are hard to understand (IMHO), so maybe these array become null when the RDG is unbound from storage.  Or maybe it was because they were initialized to nullptr.  But because Distribution assumes these arrays are not nullptr, I now initialize them in RDG's constructor.  We could change the code Distribution calls to check for nullptr, but I am always in favor of eliminating nullptr checks if we can.  In this case arrow forces us to use std::shared_ptr, but I think programmers (perhaps unwittingly) assume the member variables are initialized.

roshandathathri replied:
The same condition is checked twice inadvertently:
```cpp
else if (local_to_user_id_->length() != local_to_global_id_->length()) {
  if (local_to_user_id_->length() != local_to_global_id_->length()) {
```

witchel replied:
Superceeded by #141. 🤦

scober reviewed (approved):
```
    __    ______________  ___
   / /   / ____/_  __/  |/  /
  / /   / / __  / / / /|_/ /
 / /___/ /_/ / / / / /  / /
/_____/\____/ /_/ /_/  /_/
```
